j5/vision/markers.py

```diff
@@ -0,0 +1,203 @@
+        return Spherical(np.sqrt(self.x**2 + self.y**2 + self.z**2),
+                         np.arctan2(self.y, self.x),
+                         np.arctan2(np.sqrt(self.x**2 + self.y**2), self.z))
+
+
+class Cylindrical(NamedTuple):
+    """A Cylindrical Coordinate.
+
+    p   := axial distance
+    phi := azimuth angle (radians)
+    z   := height
+    """
+
+    p: float
+    phi: float
+    z: float
+
+    def as_cartesian(self) -> Cartesian:
+        """Provides the coordiante in the form of a cartesian coordinate."""
+        return Cartesian(self.p * np.cos(self.phi),
+                         self.p * np.sin(self.phi),
+                         self.z)
+
+    def as_cylindrical(self) -> Cylindrical:
+        """Provides the coordiante in the form of a cylindrical coordinate."""
+        return self
+
+    def as_spherical(self) -> Spherical:
+        """Provides the coordiante in the form of a spherical coordinate."""
+        return Spherical(np.sqrt(self.p**2 + self.z**2),
+                         np.arctan2(self.p, self.z),
+                         self.phi)
+
+
+class Spherical(NamedTuple):
+    """A Spherical Coordinate.
+
+    r     := radial distance
+    theta := polar angle (radians)
+    phi   := azimuth angle (radians)
+    """
+
+    r: float
+    theta: float
+    phi: float
+
+    def as_cartesian(self) -> Cartesian:
+        """Provides the coordiante in the form of a cartesian coordinate."""
+        return Cartesian(self.r * np.cos(self.phi) * np.sin(self.theta),
+                         self.r * np.sin(self.phi) * np.sin(self.theta),
+                         self.r * np.cos(self.theta))
+
+    def as_cylindrical(self) -> Cylindrical:
+        """Provides the coordiante in the form of a cylindrical coordinate."""
+        return Cylindrical(self.r * np.sin(self.theta),
+                           self.phi,
+                           self.r * np.cos(self.theta))
+
+    def as_spherical(self) -> Spherical:
+        """Provides the coordiante in the form of a spherical coordinate."""
+        return self
+
+
+Coordinate = Union[Cylindrical, Spherical, Cartesian]
+
+
+class Markers():
+    """Markers Class.
+
+    The position is stored in self.__position.
+    This is a multi-dementional numpy array.
+    The first dimesion is the marker index.
+    The second dimension is the axis.
+    Example:
+        [ [x1, y1, z1],
+          [x2, y2, z2],
+          [x3, y3, z3],
+          [x4, y4, z4] ]
+    """
+
+    def __init__(self):
+        self.__positions = None
+
+    @property
+    def positions(self) -> List[Cartesian]:
+        """Returns the makers positions."""
+        return [Cartesian(*coordinate) for coordinate in self.__positions]
+
+    def add_markers_cartesian(self, given_positions: List[Cartesian]) -> None:
+        """Add marker(s) by providing their cartesian coordinates."""
+        given_positions = np.array(given_positions)
```

Comment:
I chose to use numpy arrays because they allow the fast mutatation of the coordinate space and additional markers become less expensive (speed, not space).

Reply:
Has this been measured or is the speed increase an assumption? I'm keen to keep to the standard library as much as possible unless significant performance benefits can be demonstrated.

Reply:
Yeah, its an assumption. I will get rid of the numpy dependancy. We can look into using numpy in the future if we need to improve the performance.